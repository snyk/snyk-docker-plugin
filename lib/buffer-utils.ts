import * as crypto from "crypto";
import { HashAlgorithm } from "./types";

const HASH_ENCODING = "hex";

export async function bufferToSha1(buffer: Buffer): Promise<string> {
  const hash = crypto.createHash(HashAlgorithm.Sha1);
  const chunkSize = 100 * 1024 * 1024; // 100 MB

  return new Promise((resolve, reject) => {
    try {
      for (let offset = 0; offset < buffer.length; offset += chunkSize) {
        const end = Math.min(offset + chunkSize, buffer.length);
        const chunk = buffer.slice(offset, end);
        hash.update(chunk);
      }

      const digest = hash.digest(HASH_ENCODING);
      resolve(digest);
    } catch (err) {
      reject(err);
    }
  });
}
