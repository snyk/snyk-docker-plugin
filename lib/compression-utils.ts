import { Decompress as ZstdDecompress } from "fzstd";

/**
 * Decompresses zstd-compressed data from a buffer.
 *
 * This is a synchronous buffer-to-buffer decompression utility.
 * For streaming zstd decompression, use the decompressMaybe transform stream.
 *
 * @param compressed Buffer containing zstd-compressed data
 * @returns Decompressed data as a Buffer
 * @throws Error if decompression fails
 */
export function decompressZstd(compressed: Buffer): Buffer {
  const chunks: Buffer[] = [];

  try {
    const decompressor = new ZstdDecompress((data: Uint8Array) => {
      chunks.push(Buffer.from(data));
    });

    decompressor.push(new Uint8Array(compressed), true);

    return Buffer.concat(chunks);
  } catch (error) {
    throw new Error(
      `Zstd decompression failed: ${
        error instanceof Error ? error.message : String(error)
      }`,
    );
  }
}
