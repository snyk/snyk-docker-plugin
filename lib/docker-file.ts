import * as fs from 'fs';
import * as dockerFileParser from 'docker-file-parser';

export { getBaseImageName };

async function getBaseImageName(targetFile?: string):
  Promise<string|undefined> {

  if (!targetFile) {
    return undefined;
  }

  const contents = await readFile(targetFile);
  const commands = dockerFileParser.parse(contents);
  const fromCommands = commands.filter(command => command.name === 'FROM');
  const finalFrom = fromCommands.pop();

  return finalFrom ? String(finalFrom.args) : undefined;
}

async function readFile(path: string) {
  return new Promise((resolve, reject) => {
    fs.readFile(path, 'utf8', (err, data) => {
      return err ? reject(err) : resolve(data);
    });
  }) as Promise<string>;
}
