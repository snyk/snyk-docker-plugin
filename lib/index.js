var subProcess = require('./sub-process');
var fetchSnykDockerAnalyzer = require('./fetch-snyk-docker-analyzer');
var debug = require('debug')('snyk');

module.exports = {
  inspect: inspect,
};

function inspect(root, targetFile, options) {
  const targetImage = root;
  return fetchSnykDockerAnalyzer()
  .then(function (analyzerBinaryPath) {
    return Promise.all([
      getRuntime(),
      getDependencies(analyzerBinaryPath, targetImage),
    ])
    .then(function (result) {
      var metadata = {
        name: 'snyk-docker-plugin',
        runtime: result[0],
        packageManager: result[1].packageManager,
      };
      var package = result[1].package;
      return {
        plugin: metadata,
        package: package,
      };
    });
  })
}

function getRuntime() {
  return subProcess.execute('docker', ['version'])
  .then(function (output) {
    versionMatch = /Version:\s+(.*)\n/.exec(output);
    if (versionMatch) {
      return 'docker ' + versionMatch[1]
    }
    return undefined;
  });
}

function handleCommonErrors(error, targetImage) {
  if (error.indexOf('command not found') !== -1) {
    throw new Error('Snyk docker CLI was not found')
  }
  if (error.indexOf('Cannot connect to the Docker daemon') !== -1) {
    throw new Error('Cannot connect to the Docker daemon. Is the docker'
    + ' daemon running?')
  }
  if ((error.indexOf('Error loading image from docker engine') !== -1) ||
      (error.indexOf('Error performing image analysis') !== -1)) {
    throw new Error('Docker image was not found: ' + targetImage)
  }
  if (error.indexOf('Error getting docker client:') !== -1) {
    throw new Error('Failed getting docker client')
  }
  if (error.indexOf('Error processing image:') !== -1) {
    throw new Error('Failed processing image:' + targetImage)
  }
}

function getDependencies(analyzerBinaryPath, targetImage) {
  var result;
  return subProcess.execute(
    analyzerBinaryPath,
    buildArgs(targetImage)
  )
  .then(function (output) {
    result = parseAnalysisResults(output);
    return buildTree(
      targetImage, result.type, result.depInfosList, result.osRelease);
  })
  .then(function (package) {
    return {
      package: package,
      packageManager: result.type,
    };
  })
  .catch(function (error) {
    if (typeof error === 'string') {
      debug(`Error while running analyser: '${error}'`);
      handleCommonErrors(error, targetImage);
      errorMsg = error;
      errorMatch = /msg="(.*)"/g.exec(errorMsg)
      if (errorMatch) {
        errorMsg = errorMatch[1]
      }
      throw new Error(errorMsg);
    }

    throw error;
  })
}

function parseAnalysisResults(analysisOut) {
  var analysisJson = JSON.parse(analysisOut);

  var analysisResult = analysisJson.results.filter(function (res) {
    return res.Analysis && res.Analysis.length > 0;
  })[0];

  // TODO: if analysisResult is undefined -
  //  throw a nice error, or report 0 deps

  var depType;
  switch (analysisResult.AnalyzeType) {
    case 'Apt': {
      depType = 'deb';
      break;
    }
    default: {
      depType = analysisResult.AnalyzeType.toLowerCase()
    }
  }

  return {
    type: depType,
    depInfosList: analysisResult.Analysis,
    osRelease: analysisJson.osRelease,
  }
}

function buildTree(targetImage, depType, depInfosList, osRelease) {
  var targetSplit = targetImage.split(':');
  var imageName = targetSplit[0];
  var imageVersion = targetSplit[1] ? targetSplit[1] : 'latest';

  var root = {
    name: imageName,
    version: imageVersion,
    dockerOSRelease: osRelease,
    packageFormatVersion: depType + ':0.0.1',
    dependencies: {},
  };

  var depsMap = depInfosList.reduce(function (acc, depInfo) {
    var name = depInfo.Name;
    acc[name] = depInfo;
    return acc;
  }, {});

  var virtualDepsMap = depInfosList.reduce(function (acc, depInfo) {
    var providesNames = depInfo.Provides || [];
    providesNames.forEach(function (name) {
      acc[name] = depInfo;
    });
    return acc;
  }, {});

  var attachDeps = function (depInfos) {
    depInfos.forEach(function (depInfo) {
      var subtree = buildTreeRecurisve(
        depInfo.Name, new Set(), depsMap, virtualDepsMap);
      if (subtree) {
        root.dependencies[subtree.name] = subtree;
      }
    })
  };

  // attach (as direct deps) pkgs not marked auto-installed:
  var manuallyInstalledDeps = depInfosList.filter(function (depInfo) {
    return !depInfo.AutoInstalled;
  })
  attachDeps(manuallyInstalledDeps);

  // attach (as direct deps) pkgs marked as auto-insatalled,
  //  but not dependant upon:
  var notVisitedDeps = depInfosList.filter(function (depInfo) {
    var depName = depInfo.Name;
    return !(depsMap[depName]._visited);
  })
  attachDeps(notVisitedDeps);

  return root;
}

function buildTreeRecurisve(
    depName, ancestors, depsMap, virtualDepsMap) {
  var depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return null;
  }

  var fullName = depFullName(depInfo);
  if (ancestors.has(fullName)) {
    return null;
  }

  const tree = {
    name: fullName,
    version: depInfo.Version,
  };

  //TODO: if we do this we can avoid tracking ancestors
  if (depInfo._visited) {
    return tree;
  }
  depInfo._visited = true;

  var newAncestors = (new Set(ancestors)).add(fullName);

  var deps = depInfo.Deps || {};
  Object.keys(deps).forEach(function (depName) {
    var subTree = buildTreeRecurisve(
      depName, newAncestors, depsMap, virtualDepsMap);
    if (subTree) {
      if (!tree.dependencies) {
        tree.dependencies = {};
      }
      tree.dependencies[subTree.name] = subTree;
    }
  });

  return tree;
}

function depFullName(depInfo) {
  var fullName = depInfo.Name;
  if (depInfo.Source) {
    fullName = depInfo.Source + '/' + fullName;
  }
  return fullName;
}

function buildArgs(targetImage) {
  var args = ['analyze', targetImage];
  return args;
}
