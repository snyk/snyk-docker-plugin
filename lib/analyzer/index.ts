import * as Debug from "debug";
import { Docker, DockerOptions } from "../docker";
import * as dockerFile from "../docker-file";
import * as apkAnalyzer from "./apk-analyzer";
import * as aptAnalyzer from "./apt-analyzer";
import * as binariesAnalyzer from "./binaries-analyzer";
import * as imageInspector from "./image-inspector";
import * as osReleaseDetector from "./os-release-detector";
import * as rpmAnalyzer from "./rpm-analyzer";

export { analyze };

const debug = Debug("snyk");

const extractActions = [
  ...aptAnalyzer.APT_PKGPATHS,
  ...apkAnalyzer.APK_PKGPATHS,
  ...osReleaseDetector.OS_VERPATHS,
].map((p) => {
  return {
    name: "txt",
    pattern: p,
  };
});

async function analyze(
  targetImage: string,
  dockerfileAnalysis?: dockerFile.DockerFileAnalysis,
  options?: DockerOptions,
) {
  const docker = new Docker(targetImage, options);

  await docker.scanStaticalyIfNeeded(extractActions);

  const [imageInspection, osRelease] = await Promise.all([
    imageInspector.detect(docker),
    osReleaseDetector.detect(docker, dockerfileAnalysis),
  ]);

  const results = await Promise.all([
    apkAnalyzer.analyze(docker),
    aptAnalyzer.analyze(docker),
    rpmAnalyzer.analyze(docker),
  ]).catch((err) => {
    debug(`Error while running analyzer: '${err.stderr}'`);
    throw new Error("Failed to detect installed OS packages");
  });

  const { installedPackages, pkgManager } = getInstalledPackages(
    results as any[],
  );
  let binaries;
  try {
    binaries = await binariesAnalyzer.analyze(
      targetImage,
      installedPackages,
      pkgManager,
      options,
    );
  } catch (err) {
    debug(`Error while running binaries analyzer: '${err}'`);
    throw new Error("Failed to detect binaries versions");
  }

  return {
    imageId: imageInspection.Id,
    osRelease,
    results,
    binaries,
    imageLayers: imageInspection.RootFS && imageInspection.RootFS.Layers,
  };
}

function getInstalledPackages(
  results: any[],
): { installedPackages: string[]; pkgManager?: string } {
  const dockerAnalysis = results.find((res) => {
    return res.Analysis && res.Analysis.length > 0;
  });

  if (!dockerAnalysis) {
    return { installedPackages: [] };
  }
  const installedPackages = dockerAnalysis.Analysis.map((pkg) => pkg.Name);
  let pkgManager = dockerAnalysis.AnalyzeType;
  if (pkgManager) {
    pkgManager = pkgManager.toLowerCase();
  }
  return { installedPackages, pkgManager };
}
