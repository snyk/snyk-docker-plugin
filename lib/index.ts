const debug = require('debug')('snyk');

import * as analyzer from './analyzer';
import * as subProcess from './sub-process';
import * as dockerFile from './docker-file';

export {
  inspect,
};

function inspect(root: string, targetFile?: string, options?: any) {
  const targetImage = root;
  return Promise.all([
    getRuntime(),
    getDependencies(targetImage),
    dockerFile.getBaseImageName(targetFile),
  ])
    .then((result) => {
      const metadata = {
        name: 'snyk-docker-plugin',
        runtime: result[0],
        packageManager: result[1].packageManager,
        dockerImageId: result[1].imageId,
      };
      const pkg: any = result[1].package;
      pkg.docker = pkg.docker || {};
      pkg.docker.baseImage = result[2];
      return {
        plugin: metadata,
        package: pkg,
      };
    });
}

function getRuntime() {
  return subProcess.execute('docker', ['version'])
    .then((output) => {
      const versionMatch = /Version:\s+(.*)\n/.exec(output);
      if (versionMatch) {
        return 'docker ' + versionMatch[1];
      }
      return undefined;
    });
}

function handleCommonErrors(error, targetImage: string) {
  if (error.indexOf('command not found') !== -1) {
    throw new Error('Snyk docker CLI was not found');
  }
  if (error.indexOf('Cannot connect to the Docker daemon') !== -1) {
    throw new Error('Cannot connect to the Docker daemon. Is the docker'
    + ' daemon running?');
  }
  const ERROR_LOADING_IMAGE_STR = 'Error loading image from docker engine:';
  if (error.indexOf(ERROR_LOADING_IMAGE_STR) !== -1) {
    if (error.indexOf('reference does not exist') !== -1) {
      throw new Error(
        `Docker image was not found locally: ${targetImage}`);
    }
    if (error.indexOf('permission denied while trying to connect') !== -1) {
      let errString = error.split(ERROR_LOADING_IMAGE_STR)[1];
      errString = (errString || '').slice(0, -2); // remove trailing \"
      throw new Error(
        'Permission denied connecting to docker daemon. ' +
        'Please make sure user has the required permissions. ' +
        'Error string: ' + errString);
    }
  }
  if (error.indexOf('Error getting docker client:') !== -1) {
    throw new Error('Failed getting docker client');
  }
  if (error.indexOf('Error processing image:') !== -1) {
    throw new Error('Failed processing image:' + targetImage);
  }
}

function getDependencies(targetImage: string) {
  let result;
  return analyzer.analyze(targetImage)
    .then((output) => {
      result = parseAnalysisResults(output);
      return buildTree(
        targetImage, result.type, result.depInfosList, result.targetOS);
    })
    .then((pkg) => {
      return {
        package: pkg,
        packageManager: result.type,
        imageId: result.imageId,
      };
    })
    .catch((error) => {
      if (typeof error === 'string') {
        debug(`Error while running analyzer: '${error}'`);
        handleCommonErrors(error, targetImage);
        let errorMsg = error;
        const errorMatch = /msg="(.*)"/g.exec(errorMsg);
        if (errorMatch) {
          errorMsg = errorMatch[1];
        }
        throw new Error(errorMsg);
      }
      throw error;
    });
}

function parseAnalysisResults(analysisJson) {
  const analysisResult = analysisJson.results.filter((res) => {
    return res.Analysis && res.Analysis.length > 0;
  })[0];

  if (!analysisResult) {
    throw new Error(
      'Failed to detect a supported Linux package manager (deb/rpm/apk)');
  }

  let depType;
  switch (analysisResult.AnalyzeType) {
    case 'Apt': {
      depType = 'deb';
      break;
    }
    default: {
      depType = analysisResult.AnalyzeType.toLowerCase();
    }
  }

  return {
    imageId: analysisJson.imageId,
    targetOS: analysisJson.osRelease,
    type: depType,
    depInfosList: analysisResult.Analysis,
  };
}

function buildTree(targetImage, depType, depInfosList, targetOS) {
  const targetSplit = targetImage.split(':');
  const imageName = targetSplit[0];
  const imageVersion = targetSplit[1] ? targetSplit[1] : 'latest';

  const root = {
    name: imageName,
    version: imageVersion,
    targetOS,
    packageFormatVersion: depType + ':0.0.1',
    dependencies: {},
  };

  const depsMap = depInfosList.reduce((acc, depInfo) => {
    const name = depInfo.Name;
    acc[name] = depInfo;
    return acc;
  }, {});

  const virtualDepsMap = depInfosList.reduce((acc, depInfo) => {
    const providesNames = depInfo.Provides || [];
    providesNames.forEach((name) => {
      acc[name] = depInfo;
    });
    return acc;
  }, {});

  const depsCounts = {};
  depInfosList.forEach((depInfo) => {
    countDepsRecursive(
      depInfo.Name, new Set(), depsMap, virtualDepsMap, depsCounts);
  });
  const DEP_FREQ_THRESHOLD = 100;
  const tooFrequentDepNames = Object.keys(depsCounts)
    .filter((depName) => {
      return depsCounts[depName] > DEP_FREQ_THRESHOLD;
    });

  const attachDeps = (depInfos) => {
    const depNamesToSkip = new Set(tooFrequentDepNames);
    depInfos.forEach((depInfo) => {
      const subtree = buildTreeRecurisve(
        depInfo.Name, new Set(), depsMap, virtualDepsMap, depNamesToSkip);
      if (subtree) {
        root.dependencies[subtree.name] = subtree;
      }
    });
  };

  // attach (as direct deps) pkgs not marked auto-installed:
  const manuallyInstalledDeps = depInfosList.filter((depInfo) => {
    return !depInfo.AutoInstalled;
  });
  attachDeps(manuallyInstalledDeps);

  // attach (as direct deps) pkgs marked as auto-insatalled,
  //  but not dependant upon:
  const notVisitedDeps = depInfosList.filter((depInfo) => {
    const depName = depInfo.Name;
    return !(depsMap[depName]._visited);
  });
  attachDeps(notVisitedDeps);

  // group all the "too frequest" deps under a meta package:
  if (tooFrequentDepNames.length > 0) {
    const tooFrequentDeps = tooFrequentDepNames.map((name) => {
      return depsMap[name];
    });

    const metaSubtree = {
      name: 'meta-common-packages',
      version: 'meta',
      dependencies: {},
    };

    tooFrequentDeps.forEach((depInfo) => {
      const pkg = {
        name: depFullName(depInfo),
        version: depInfo.Version,
      };
      metaSubtree.dependencies[pkg.name] = pkg;
    });

    root.dependencies[metaSubtree.name] = metaSubtree;
  }

  return root;
}

function buildTreeRecurisve(
  depName, ancestors, depsMap, virtualDepsMap, depNamesToSkip) {
  const depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return null;
  }

  // "realName" as the argument depName might be a virtual pkg
  const realName = depInfo.Name;
  const fullName = depFullName(depInfo);
  if (ancestors.has(fullName) || depNamesToSkip.has(realName)) {
    return null;
  }

  const tree: {
    name: string;
    version: string;
    dependencies?: any;
  } = {
    name: fullName,
    version: depInfo.Version,
  };

  if (depInfo._visited) {
    return tree;
  }
  depInfo._visited = true;

  const newAncestors = (new Set(ancestors)).add(fullName);

  const deps = depInfo.Deps || {};
  Object.keys(deps).forEach((name) => {
    const subTree = buildTreeRecurisve(
      name, newAncestors, depsMap, virtualDepsMap, depNamesToSkip);
    if (subTree) {
      if (!tree.dependencies) {
        tree.dependencies = {};
      }
      tree.dependencies[subTree.name] = subTree;
    }
  });

  return tree;
}

function countDepsRecursive(
  depName, ancestors, depsMap, virtualDepsMap, depCounts) {
  const depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return;
  }

  // "realName" as the argument depName might be a virtual pkg
  const realName = depInfo.Name;
  if (ancestors.has(realName)) {
    return;
  }

  depCounts[realName] = (depCounts[realName] || 0) + 1;

  const newAncestors = (new Set(ancestors)).add(realName);
  const deps = depInfo.Deps || {};
  Object.keys(deps).forEach((name) => {
    countDepsRecursive(
      name, newAncestors, depsMap, virtualDepsMap, depCounts);
  });
}

function depFullName(depInfo) {
  let fullName = depInfo.Name;
  if (depInfo.Source) {
    fullName = depInfo.Source + '/' + fullName;
  }
  return fullName;
}
