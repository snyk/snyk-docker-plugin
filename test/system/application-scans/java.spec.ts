import { scan } from "../../../lib";
import { getFixture } from "../../util";

describe("jar binaries scanning", () => {
  it("should return expected result", async () => {
    const fixturePath = getFixture("docker-archives/docker-save/java.tar");
    const imageNameAndTag = `docker-archive:${fixturePath}`;

    const pluginResult = await scan({
      path: imageNameAndTag,
      "app-vulns": true,
    });

    expect(pluginResult).toMatchSnapshot();
  });

  describe("uber jars", () => {
    let fixturePath;
    let pluginResult;
    let fingerprints;

    const fatJar = {
      location: "/uberjar.jar",
      digest: expect.any(String),
      dependencies: expect.any(Array),
    };
    const nestedJar = {
      location: "/uberjar.jar/lib/guava-30.1-jre.jar",
      digest: null,
      artifactId: "guava",
      groupId: "com.google.guava",
      version: "30.1-jre",
      dependencies: expect.any(Array),
    };

    // TODO: deprecate --shaded-jars-depth and leave only --nested-jars-depth
    describe.each(["shaded-jars-depth", "nested-jars-depth"])(
      "--%s",
      (flagName) => {
        describe(`with all needed CLI flags (app-vulns and ${flagName})`, () => {
          beforeAll(async () => {
            // Arrange
            fixturePath = getFixture(
              "docker-archives/docker-save/java-uberjar.tar",
            );
            const imageNameAndTag = `docker-archive:${fixturePath}`;

            // Act
            pluginResult = await scan({
              path: imageNameAndTag,
              "app-vulns": true,
              [flagName]: "1",
            });

            fingerprints =
              pluginResult.scanResults[1].facts[0].data.fingerprints;
          });

          it("should return the Uber JAR and two nested JARs in the results", () => {
            expect(fingerprints).toHaveLength(3);
          });

          it("should return nested (second-level) jar in the result", () => {
            expect(fingerprints).toContainEqual(nestedJar);
          });

          it("should return a first-level jar that have nested jars in it (uber jar)", () => {
            expect(fingerprints).toContainEqual(fatJar);
          });

          it("should be a unique set of fingerprints", () => {
            const isUniqueList = (arr: any[]): boolean =>
              arr.length === new Set(arr).size;
            expect(isUniqueList(fingerprints)).toBeTruthy();
          });

          it("should return first-level jars that have no nested jars in it", () => {
            expect(fingerprints).toContainEqual({
              location: "/j2objc-annotations-1.3.jar",
              digest: null,
              dependencies: expect.any(Array),
              artifactId: "j2objc-annotations",
              groupId: "com.google.j2objc",
              version: "1.3",
            });
          });
        });

        describe(`with default ${flagName}`, () => {
          // Arrange
          let imageNameAndTag;
          beforeAll(async () => {
            fixturePath = getFixture(
              "docker-archives/docker-save/java-uberjar.tar",
            );
            imageNameAndTag = `docker-archive:${fixturePath}`;
          });

          it("should return nested (second-level) jar in the result", async () => {
            // Act
            pluginResult = await scan({
              path: imageNameAndTag,
              "app-vulns": true,
              [flagName]: true,
            });

            // Assert
            fingerprints =
              pluginResult.scanResults[1].facts[0].data.fingerprints;

            expect(fingerprints).toContainEqual(nestedJar);
          });

          it(`should unpack 1 level of jars if ${flagName} flag is missing`, async () => {
            // Act
            pluginResult = await scan({
              path: imageNameAndTag,
              "app-vulns": true,
            });

            fingerprints =
              pluginResult.scanResults[1].facts[0].data.fingerprints;

            expect(fingerprints).toContainEqual(nestedJar);
          });
        });

        describe("with missing CLI flags", () => {
          fixturePath = getFixture(
            "docker-archives/docker-save/java-uberjar.tar",
          );
          const imageNameAndTag = `docker-archive:${fixturePath}`;

          it(`should not unpack jars if ${flagName} flag is set to 0`, async () => {
            // Act
            pluginResult = await scan({
              path: imageNameAndTag,
              "app-vulns": true,
              [flagName]: "0",
            });

            // Assert
            fingerprints =
              pluginResult.scanResults[1].facts[0].data.fingerprints;
            expect(fingerprints).toContainEqual(fatJar);
            expect(fingerprints).not.toContainEqual(nestedJar);
          });

          it(`should throw if ${flagName} flag is set to -1`, async () => {
            // Act + Assert
            await expect(
              scan({
                path: imageNameAndTag,
                "app-vulns": true,
                [flagName]: "-1",
              }),
            ).rejects.toThrow();
          });

          it("should throw error if app-vulns flag is missing", async () => {
            // Act
            await expect(
              scan({
                path: imageNameAndTag,
                [flagName]: "1",
              }),
            ).rejects.toThrow();
          });

          it(`should throw error if ${flagName} is not a number`, async () => {
            // Act
            await expect(
              scan({
                path: imageNameAndTag,
                "app-vulns": true,
                [flagName]: "NotANumber!",
              }),
            ).rejects.toThrow();
          });
        });

        describe("multi-level jars", () => {
          let imageNameAndTag;
          beforeAll(async () => {
            // Arrange
            fixturePath = getFixture(
              "docker-archives/docker-save/3-level-jar.tar",
            );
            imageNameAndTag = `docker-archive:${fixturePath}`;
          });

          it(`should return partial scan if ${flagName}=1`, async () => {
            const level2JarFingerprint = {
              location: "/level-3-jar.jar/level-2-jar.jar",
              digest: expect.any(String),
            };

            // Act
            pluginResult = await scan({
              path: imageNameAndTag,
              "app-vulns": true,
              [flagName]: "1",
            });

            fingerprints =
              pluginResult.scanResults[1].facts[0].data.fingerprints;

            expect(fingerprints).toHaveLength(2);
            expect(fingerprints).toContainEqual(
              expect.objectContaining(level2JarFingerprint),
            );
          });

          it(`should return full scan if ${flagName}=2, because unpacking 2 levels will reveal the third`, async () => {
            const deepestLevelJarFingerprint = {
              location:
                "/level-3-jar.jar/level-2-jar.jar/lib/listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar",
              digest: null,
              dependencies: expect.any(Array),
              artifactId: "listenablefuture",
              groupId: "com.google.guava",
              version: "9999.0-empty-to-avoid-conflict-with-guava",
            };

            // Act
            pluginResult = await scan({
              path: imageNameAndTag,
              "app-vulns": true,
              [flagName]: "2",
            });

            fingerprints =
              pluginResult.scanResults[1].facts[0].data.fingerprints;

            expect(fingerprints).toHaveLength(3);
            expect(fingerprints).toContainEqual(deepestLevelJarFingerprint);
          });

          it(`should return full scan if ${flagName}=4, because specifiying more levels than exist should not break things`, async () => {
            const deepestLevelJarFingerprint = {
              location:
                "/level-3-jar.jar/level-2-jar.jar/lib/listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar",
              digest: null,
              dependencies: expect.any(Array),
              artifactId: "listenablefuture",
              groupId: "com.google.guava",
              version: "9999.0-empty-to-avoid-conflict-with-guava",
            };

            // Act
            pluginResult = await scan({
              path: imageNameAndTag,
              "app-vulns": true,
              [flagName]: "4",
            });

            fingerprints =
              pluginResult.scanResults[1].facts[0].data.fingerprints;

            expect(fingerprints).toHaveLength(3);
            expect(fingerprints).toContainEqual(deepestLevelJarFingerprint);
          });

          it("should handle sibling uber jars", async () => {
            // Arrange
            fixturePath = getFixture(
              "docker-archives/docker-save/sibling-uberjars-deepest-first.tar",
            );
            imageNameAndTag = `docker-archive:${fixturePath}`;
            const threeLevelFingerprint = {
              location:
                "/A-level-3-jar.jar/level-2-jar.jar/lib/listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar",
              digest: null,
              dependencies: expect.any(Array),
              artifactId: "listenablefuture",
              groupId: "com.google.guava",
              version: "9999.0-empty-to-avoid-conflict-with-guava",
            };
            const twoLevelFingerprint = {
              location: "/B-uber-jar.jar/guava-30.1-jre.jar",
              digest: null,
              dependencies: expect.any(Array),
              artifactId: "guava",
              groupId: "com.google.guava",
              version: "30.1-jre",
            };
            const flatFingerprint = {
              location: "/C-j2objc-annotations-1.3.jar",
              digest: null,
              dependencies: expect.any(Array),
              artifactId: "j2objc-annotations",
              groupId: "com.google.j2objc",
              version: "1.3",
            };

            // Act
            pluginResult = await scan({
              path: imageNameAndTag,
              "app-vulns": true,
              [flagName]: "4",
            });

            fingerprints =
              pluginResult.scanResults[1].facts[0].data.fingerprints;

            // Assert
            expect(fingerprints).toHaveLength(6);
            expect(fingerprints).toContainEqual(threeLevelFingerprint);
            expect(fingerprints).toContainEqual(twoLevelFingerprint);
            expect(fingerprints).toContainEqual(flatFingerprint);
          });

          // TODO CAP-447
          it.skip("should return correct levels unpacked for sibling jars, where the last is the deepest", async () => {
            // Arrange
            fixturePath = getFixture(
              "docker-archives/docker-save/sibling-uberjars-shallowest-first.tar",
            );
            imageNameAndTag = `docker-archive:${fixturePath}`;
          });

          // TODO CAP-447
          it.skip("should return correct levels unpacked for sibling jars, where the first is the deepest", async () => {
            // Arrange
            fixturePath = getFixture(
              "docker-archives/docker-save/sibling-uberjars-deepest-first.tar",
            );
            imageNameAndTag = `docker-archive:${fixturePath}`;

            // Act
            pluginResult = await scan({
              path: imageNameAndTag,
              "app-vulns": true,
              [flagName]: "2",
            });

            fingerprints =
              pluginResult.scanResults[1].facts[0].data.fingerprints;
            // tslint:disable-next-line:no-console
            console.log("🚀 ~ fingerprints", fingerprints);
          });

          it("should return correct result for top level jar that contains 2 uberjars within it", async () => {
            // Arrange
            fixturePath = getFixture(
              "docker-archives/docker-save/top-level.tar",
            );
            imageNameAndTag = `docker-archive:${fixturePath}`;

            const firstSibling = {
              location:
                "/top-level.jar/A-level-3-jar.jar/level-2-jar.jar/lib/listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar",
              digest: null,
              dependencies: expect.any(Array),
              artifactId: "listenablefuture",
              groupId: "com.google.guava",
              version: "9999.0-empty-to-avoid-conflict-with-guava",
            };
            const secondSibling = {
              location: "/top-level.jar/B-uber-jar.jar/guava-30.1-jre.jar",
              digest: null,
              dependencies: expect.any(Array),
              artifactId: "guava",
              groupId: "com.google.guava",
              version: "30.1-jre",
            };

            // Act
            pluginResult = await scan({
              path: imageNameAndTag,
              "app-vulns": true,
              [flagName]: "4",
            });

            fingerprints =
              pluginResult.scanResults[1].facts[0].data.fingerprints;

            // Assert
            expect(fingerprints).toHaveLength(6);
            expect(fingerprints).toContainEqual(firstSibling);
            expect(fingerprints).toContainEqual(secondSibling);
          });
        });
      },
    );
  });
});
