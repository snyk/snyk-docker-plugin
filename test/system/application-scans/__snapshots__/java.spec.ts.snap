// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jar binaries scanning should return expected result 1`] = `
Object {
  "scanResults": Array [
    Object {
      "facts": Array [
        Object {
          "data": Object {
            "graph": Object {
              "nodes": Array [
                Object {
                  "deps": Array [],
                  "nodeId": "root-node",
                  "pkgId": "docker-image|java.tar@",
                },
              ],
              "rootNodeId": "root-node",
            },
            "pkgManager": Object {
              "name": "linux",
              "repositories": Array [
                Object {
                  "alias": "unknown:0.0",
                },
              ],
            },
            "pkgs": Array [
              Object {
                "id": "docker-image|java.tar@",
                "info": Object {
                  "name": "docker-image|java.tar",
                  "version": undefined,
                },
              },
            ],
            "schemaVersion": "1.3.0",
          },
          "type": "depGraph",
        },
        Object {
          "data": "sha256:e997a24cd4633de60213a23debcfb37dcc9fa6578a112976fc80aaa094ea1f59",
          "type": "imageId",
        },
        Object {
          "data": Array [
            "d446fd2d006bff0982827588b2b613f91a4e2a59d619da68fa7ce1c880b8d907/layer.tar",
          ],
          "type": "imageLayers",
        },
        Object {
          "data": "2020-03-24T13:05:21.8612979Z",
          "type": "imageCreationTime",
        },
        Object {
          "data": Array [
            "sha256:05356880e40587a7dedbdcb39f8a9f99f2f48e094c7e1d2fae64f4774cc7a4aa",
          ],
          "type": "rootFs",
        },
      ],
      "identity": Object {
        "args": Object {
          "platform": "linux/amd64",
        },
        "type": "linux",
      },
      "target": Object {
        "image": "docker-image|java.tar",
      },
    },
    Object {
      "facts": Array [
        Object {
          "data": Object {
            "fingerprints": Array [
              Object {
                "dependencies": Array [],
                "digest": "485de3a253e23f645037828c07f1d7f1af40763a",
                "location": "/app/activation-1.1.1.jar",
              },
            ],
            "origin": "java.tar",
            "path": "/app",
          },
          "type": "jarFingerprints",
        },
        Object {
          "data": "sha256:e997a24cd4633de60213a23debcfb37dcc9fa6578a112976fc80aaa094ea1f59",
          "type": "imageId",
        },
      ],
      "identity": Object {
        "targetFile": "/app",
        "type": "maven",
      },
      "target": Object {
        "image": "docker-image|java.tar",
      },
    },
    Object {
      "facts": Array [
        Object {
          "data": Object {
            "fingerprints": Array [
              Object {
                "artifactId": "aopalliance-repackaged",
                "dependencies": Array [],
                "digest": null,
                "groupId": "org.glassfish.hk2.external",
                "location": "/libs/aopalliance-repackaged-2.5.0.jar",
                "version": "2.5.0",
              },
              Object {
                "artifactId": "audience-annotations",
                "dependencies": Array [],
                "digest": null,
                "groupId": "org.apache.yetus",
                "location": "/libs/audience-annotations-0.5.0.jar",
                "version": "0.5.0",
              },
            ],
            "origin": "java.tar",
            "path": "/libs",
          },
          "type": "jarFingerprints",
        },
        Object {
          "data": "sha256:e997a24cd4633de60213a23debcfb37dcc9fa6578a112976fc80aaa094ea1f59",
          "type": "imageId",
        },
      ],
      "identity": Object {
        "targetFile": "/libs",
        "type": "maven",
      },
      "target": Object {
        "image": "docker-image|java.tar",
      },
    },
  ],
}
`;
